### ISSUE TITLE
Oneway RPC support

### ISSUE THREAD

**Issue by breerly** [yarpc/yarpc-go#327]
Currently, we only support Unary RPCs. In order to finalize our APIs, we need to support at least 2.

Let's add Oneway support to all our encodings over our existing transports: HTTP & TChannel.

**Comment by AlexAPB** [comment]
**Proposal**

**Encoding**
- thrift/protobufs/etc need to have some way to define which procedures are of type Oneway

**Config**
- yaml files contain some way (TBD) of defining multiple outbound RPC types (unary, oneway, etc) per service
    - this allows a service to choose which transport it will use to call a specific procedure
        - eg: 
        - calling ServiceA::doSomething() can use a kafka transport
        - calling ServiceA::doSomethingElse() can use http
        - calling ServiceA::doSomethingDifferent() can use kafka configured a different way (eg different topics)

**Dispatcher/Channels**
- Dispatcher holds `map[service]map[procedure]Outbound` (instead of an array of multiple service Outbounds)
- `Dispatcher.Channel(‘service-name’)` would return a channel backed by multiple Outbounds and hold `map[procedure]Outbound`
-` channel.GetOutbound()` will take in a procedure name so we can get the corresponding Outbound type

**Outbounds**
- there will be different types of Outbounds as return types will differ:

```
type OneWayOutbound interface {
	Outbound
	Call(ctx context.Context, request *Request) (*Ack, error)
}
type UnaryOutbound interface {
	Outbound
	Call(ctx context.Context, request *Request) (*Response, error)
}
type StreamingOutbound interface {
…
```

encodings/[thrift | json | raw]/outbounds.go files will change their client interface to hold other Call types, eg:
```
type Client interface {
	CallUnary(ctx context.Context, reqMeta yarpc.CallReqMeta, reqBody envelope.Enveloper) (wire.Value, yarpc.CallResMeta, error)
	CallOneway(ctx context.Context, reqMeta yarpc.CallReqMeta, reqBody envelope.Enveloper) (yarpc.ack, yarpc.CallResMeta, error)
	CallStreaming(ctx …
}

```
- within these functions they’ll call the following to get the appropriate outbound
```
	out := client.channel.GetOutbound(‘procedure-name’)
	…
	out.Call(ctx,req)
```

**Clients**
- generated client code for oneway procedures will need to have return types of `(yarpc.Ack,err)`
- these generated functions will call the corresponding CallUnary/Oneway/etc in outbounds.go (as mentioned above)

**Transport/outbounds.go**
- TChannel/Http will need to implement each of the Outbound interfaces it implements (eg, OnewayOutobund and UnaryOutbound)


**Inbounds**
- We’ll create Handler interfaces for each RPC type
```
type UnaryHandler interface {
	Handle(
		ctx context.Context,
		opts Options,
		req *Request,
		resw ResponseWriter,
	) error
}

// OneWayHandler ...
type OneWayHandler interface {
	Handle(
		ctx context.Context,
		opts Options,
		req *Request,
	) error
}
```

**Other Notes/Caveats**
- filters/interceptors will need to be RPC type specific since each RPC type has a different function signature
- there are a lot of config changes that will need to be changed to support the multiple outbounds which looks like it would modify xyaprc-go quite a bit
- there are of course a few things that still need to be ironed out, but I thought I'd share some insight on the direction I'm going towards
- all feedback is welcome :)

**Comment by abhinav** [comment]
Sounds reasonable. I would like to note, though, that we may want to rename the `Call` and `Handle` methods for the different transport types to `CallUnary`, `CallOneWay`, etc. so that they can all be implemented by the same object.

**Comment by breerly** [comment]
Looking good to me - make the changes @abhinav suggested, and then I'll rope Xu for feedback :)

**Comment by AlexAPB** [comment]
Updated proposal to specify that we will not rename any unary functions throughout since unary is the expected default.

**Comment by AlexAPB** [comment]
resolved #420
